src/darwin/Framework/CHIP/CHIPDevice.mm

```diff
@@ -173,6 +202,898 @@ - (void)subscribeWithQueue:(dispatch_queue_t)queue
+    OnDoneCallbackType mOnDone;
+    OnSubscriptionEstablishedCallbackType mOnSubscriptionEstablished;
+    app::BufferedReadCallback mBufferedReadAdapter;
+    Platform::UniquePtr<app::ReadClient> mReadClient;
+};
+
+- (void)readAttributeWithEndpointId:(NSUInteger)endpointId
+                          clusterId:(NSUInteger)clusterId
+                        attributeId:(NSUInteger)attributeId
+                        clientQueue:(dispatch_queue_t)clientQueue
+                         completion:(CHIPDeviceResponseHandler)completion
+{
+    new NSObjectAttributeCallbackBridge(
+        clientQueue, completion, ^(chip::Callback::Cancelable * success, chip::Callback::Cancelable * failure) {
+            auto successFn = chip::Callback::Callback<NSObjectAttributeCallback>::FromCancelable(success);
+            auto failureFn = chip::Callback::Callback<CHIPErrorCallback>::FromCancelable(failure);
+            auto context = successFn->mContext;
+            auto successCb = successFn->mCall;
+            auto failureCb = failureFn->mCall;
+            auto resultArray = [[NSMutableArray alloc] init];
+            auto resultSuccess = [[NSMutableArray alloc] init];
+            auto resultFailure = [[NSMutableArray alloc] init];
+            auto onSuccessCb = [resultArray, resultSuccess](
+                                   const app::ConcreteAttributePath & attribPath, const NSObjectData & aData) {
+                [resultArray
+                    addObject:[[NSDictionary alloc] initWithObjectsAndKeys:[NSNumber numberWithUnsignedInt:attribPath.mEndpointId],
+                                                    kCHIPEndpointIdKey, [NSNumber numberWithUnsignedInt:attribPath.mClusterId],
+                                                    kCHIPClusterIdKey, [NSNumber numberWithUnsignedInt:attribPath.mAttributeId],
+                                                    kCHIPAttributeIdKey, aData.GetDecodedObject(), kCHIPDataKey,
+                                                    [NSNumber numberWithUnsignedInt:0], kCHIPStatusKey, nil]];
+                if ([resultSuccess count] == 0) {
+                    [resultSuccess addObject:[NSNumber numberWithBool:YES]];
+                }
+            };
+
+            auto onFailureCb = [resultArray, resultFailure](const app::ConcreteAttributePath * attribPath, CHIP_ERROR aError) {
+                if (attribPath) {
+                    [resultArray addObject:[[NSDictionary alloc]
+                                               initWithObjectsAndKeys:[NSNumber numberWithUnsignedInt:attribPath->mEndpointId],
+                                               kCHIPEndpointIdKey, [NSNumber numberWithUnsignedInt:attribPath->mClusterId],
+                                               kCHIPClusterIdKey, [NSNumber numberWithUnsignedInt:attribPath->mAttributeId],
+                                               kCHIPAttributeIdKey, [NSNumber numberWithUnsignedInteger:aError.AsInteger()],
```

Comment:
That AsInteger() thing does not have a stable encoding; it's subject to change.  We don't want to be exposing that as part of our API surface.

And even if it does not change, extracting the actual error from it requires knowledge that is not defined in our API....

Reply:
Gotcha. Would CHIPError be always compatible or can it change its value semantics? We cannot assume that the iOS and primary resident have the same version at all. Hence, if CHIPError code itself could change, instead of using NSError derived from CHIPError as suggested previously, we should just add a key "failed" with value true for all failure cases and potentially add another key "errorMsg" for an error string.

Reply:
> Would CHIPError be always compatible or can it change its value semantics?

I'm not 100% sure what you're asking, but here's my best understanding of the stability of this stuff:

1. CHIPError encoding interaction model status.  The actual interaction model status will generally not change (since it's defined by the spec), but the way it's stuffed into a CHIPError might.  If we convert to our NSError representation, that will just reflect the spec and not depend on the specific way CHIPError encodes spec-defined errors.
2. CHIPError encoding some sort of SDK-internal error code.  These are not stable at all.  Our mapping to NSError onlt reflects some of them, and we could probably at least make sure that we don't start reusing the same `code` with a different meaning there, so the NSError would likely be more stable than the CHIPError integer.

We could do the boolean "failed" with an error string, but I think the NSError representation, if it's not too much pain to marshal, actually gives us what we want stability-wise.